### PR TITLE
Fix WebUI API doesn't update IP and Port

### PR DESCRIPTION
**Pull request by myfingerhurt**
Add a checker which prevents of adding your own gpg_id.
Add an inactive friends list based on last seen Unix timestamp.
Add a method of removing a specific peer based on peer_id.

### CODE REVIEW

**Comment by csoler** [comment]
what do you mean by "not adding your own PGP Id" ? You should be able to add yourself as a friend when you have multiple locations.

**Comment by myfingerhurt** [comment]
As fas as I know, the same pgp_id will discorvery themself and add online subnodes automaticlly even you remove one of them their just keep coming back, but if you accidently add your own peer_id on the WebUI, the peer list will show two identical PGP names and its subnodes, untill the retroshare gets restarted.  

And this modification don't add the same peer_id as long as it belongs to your current node, it doesn't mean you can't add your new nodes with the same pgp_id but different peer_id.

File: libresapi/src/api/PeersHandler.cpp [hunk 1]
```diff
@@ -691,6 +792,107 @@ void PeersHandler::handleExamineCert(Request &req, Response &resp)
...
+/**
+ *  \brief Remove specific location or user
+ *  
+ *  \param [in] req request from user either peer_id or gpg_id is needed.
+ *  \param [in] resp response to user
+ *  \return 
+ *  
+ *  \details 
+ */
+void PeersHandler::handleRemoveNode(Request &req, Response &resp)
```
Comment (G10h4ck) [inline libresapi/src/api/PeersHandler.cpp:817]: Don't like this approach, the function should work with location or with PGP id only, having so different behaviour depending on how parameters are passed is confusing, more over there is already another way to remove a PGP friend via API and it's 

/peers/${PGP_ID}/delete

So in my opinion this function should be for location only

File: libresapi/src/api/PeersHandler.cpp [hunk 2]
```diff
@@ -691,6 +792,107 @@ void PeersHandler::handleExamineCert(Request &req, Response &resp)
...
+ *  \brief Retrieve inactive user list before specific time
+ *  
+ *  \param [in] req request from user, datatime.
+ *  \param [in] resp response to request
+ *  \return a pgp_id list.
+ *  
+ *  \details Combine this with handleRemoveNode, 
+ *  batch remove inactive users are possible and convenient.
+ */
+void PeersHandler::handleGetInactiveUsers(Request &req, Response &resp)
```
Comment (G10h4ck) [inline libresapi/src/api/PeersHandler.cpp:854]: As the API is used as a gateway to other languages too that may have different way of representing the date the representation format should be specified in the documentation

File: libresapi/src/api/PeersHandler.cpp [hunk 3]
```diff
@@ -624,34 +632,127 @@ void PeersHandler::handleWildcard(Request &req, Response &resp)
...
+			//RsPeerId peer_id;
+			//RsPgpId pgp_id;
+			//std::string cleanCert;
+			uint32_t error_code;
+			std::string error_string;
+			RsPeerDetails peerDetails;
+			//std::string debug_string = " ";
+			RsPgpId own_pgp = mRsPeers->getGPGOwnId();
+			RsPeerId ownpeer_id = mRsPeers->getOwnId();
+			do{
```
Comment (G10h4ck) [inline libresapi/src/api/PeersHandler.cpp]: please use Allman indent style as per http://retroshare.readthedocs.io/en/latest/developer/coding/#coding-style

File: libresapi/src/api/PeersHandler.cpp [hunk 4]
```diff
@@ -624,34 +632,127 @@ void PeersHandler::handleWildcard(Request &req, Response &resp)
...
+					{
+						std::cerr << "handleWildcard addPeer: is own peer_id, ignore" << std::endl;
+						resp.mDebug << "Error: failed to add peer_id, because its your own peer_id" << std::endl;
+						resp.mDebug << error_string << std::endl;
+						break;
+					}
+					
+					std::cerr << "handleWildcard addPeer: is own pgp_id, update only" << std::endl;
+					
+				}else{
```
Comment (G10h4ck) [inline libresapi/src/api/PeersHandler.cpp]: Please use Allman indent style http://retroshare.readthedocs.io/en/latest/developer/coding/#coding-style

File: libresapi/src/api/PeersHandler.cpp [hunk 5]
```diff
@@ -624,34 +632,127 @@ void PeersHandler::handleWildcard(Request &req, Response &resp)
...
+			   }
+				
+				ok = true;
+				
+				// Retrun node details to response, after succes of adding peers
+				peerDetailsToStream(resp.mDataStream, peerDetails);
+				
+				std::cerr << "handleWildcard addPeer: peerDetailsToStream" << std::endl;
+				
+				if (!peerDetails.location.empty()) {
```
Comment (G10h4ck) [inline libresapi/src/api/PeersHandler.cpp]: Please use Allman indent style http://retroshare.readthedocs.io/en/latest/developer/coding/#coding-style

File: libresapi/src/api/PeersHandler.cpp [hunk 6]
```diff
@@ -624,34 +632,127 @@ void PeersHandler::handleWildcard(Request &req, Response &resp)
...
+						std::cerr << std::endl;
+						mRsPeers->setExtAddress(peerDetails.id, peerDetails.extAddr, peerDetails.extPort);
+					}
+					if (!peerDetails.localAddr.empty() && peerDetails.localPort) {
+						std::cerr << "handleWildcard addPeer: setting ip local address." ;
+						std::cerr << peerDetails.localAddr;
+						std::cerr << " Port: " << peerDetails.localPort << std::endl;
+						mRsPeers->setLocalAddress(peerDetails.id, peerDetails.localAddr, peerDetails.localPort);
+					}
+					if (!peerDetails.dyndns.empty()) {
```
Comment (G10h4ck) [inline libresapi/src/api/PeersHandler.cpp]: Please use Allman indent style http://retroshare.readthedocs.io/en/latest/developer/coding/#coding-style

File: libresapi/src/api/PeersHandler.cpp [hunk 7]
```diff
@@ -691,6 +792,107 @@ void PeersHandler::handleExamineCert(Request &req, Response &resp)
...
+void PeersHandler::handleRemoveNode(Request &req, Response &resp)
+{
+	std::string ssl_peer_id;
+	std::string gpg_id;
+	req.mStream << makeKeyValueReference("peer_id", ssl_peer_id);
+	req.mStream << makeKeyValueReference("gpg_id", gpg_id);
+	RsPeerId peerId(ssl_peer_id);
+	RsPgpId gpgId(gpg_id);
+	
+	if(!gpgId.isNull()){
```
Comment (G10h4ck) [inline libresapi/src/api/PeersHandler.cpp]: Please use Allman indent style http://retroshare.readthedocs.io/en/latest/developer/coding/#coding-style

File: libresapi/src/api/PeersHandler.cpp [hunk 8]
```diff
@@ -691,6 +792,107 @@ void PeersHandler::handleExamineCert(Request &req, Response &resp)
...
+		mRsPeers->removeFriend(gpgId);
+		resp.mDataStream << makeKeyValue("gpg_id", gpg_id);
+		resp.setOk(); 
+	}else if(!peerId.isNull()){
```
Comment (G10h4ck) [inline libresapi/src/api/PeersHandler.cpp]: Please use Allman indent style http://retroshare.readthedocs.io/en/latest/developer/coding/#coding-style

File: libresapi/src/api/PeersHandler.cpp [hunk 9]
```diff
@@ -691,6 +792,107 @@ void PeersHandler::handleExamineCert(Request &req, Response &resp)
...
+		mRsPeers->removeFriendLocation(peerId);
+		resp.mDataStream << makeKeyValue("peer_id", ssl_peer_id);
+		resp.setOk();        
+	}else{
```
Comment (G10h4ck) [inline libresapi/src/api/PeersHandler.cpp]: Please use Allman indent style http://retroshare.readthedocs.io/en/latest/developer/coding/#coding-style

File: libresapi/src/api/PeersHandler.cpp [hunk 10]
```diff
@@ -691,6 +792,107 @@ void PeersHandler::handleExamineCert(Request &req, Response &resp)
...
+ *  batch remove inactive users are possible and convenient.
+ */
+void PeersHandler::handleGetInactiveUsers(Request &req, Response &resp)
+{
+	std::string datetime;
+	req.mStream << makeKeyValueReference("datetime", datetime);
+	
+	uint32_t before = strtoul(datetime.c_str(), NULL, 10);
+	
+	if(datetime.empty()){
```
Comment (G10h4ck) [inline libresapi/src/api/PeersHandler.cpp]: Please use Allman indent style http://retroshare.readthedocs.io/en/latest/developer/coding/#coding-style

File: libresapi/src/api/PeersHandler.cpp [hunk 11]
```diff
@@ -691,6 +792,107 @@ void PeersHandler::handleExamineCert(Request &req, Response &resp)
...
+	// Make pair of gpg_id and last connect time
+	std::map<RsPgpId, uint32_t> mapRsPgpId;
+	
+	// get all peers' infomations
+	for(std::list<RsPeerId>::iterator lt = peers.begin(); lt != peers.end(); ++lt)
+	{
+		RsPeerDetails details;
+		mRsPeers->getPeerDetails(*lt, details);
+		// Search for exist pgp_id, update the last connect time
+		if(mapRsPgpId.find(details.gpg_id) != mapRsPgpId.end()){
```
Comment (G10h4ck) [inline libresapi/src/api/PeersHandler.cpp]: Please use Allman indent style http://retroshare.readthedocs.io/en/latest/developer/coding/#coding-style

File: libresapi/src/api/PeersHandler.cpp [hunk 12]
```diff
@@ -691,6 +792,107 @@ void PeersHandler::handleExamineCert(Request &req, Response &resp)
...
+			if(mapRsPgpId[details.gpg_id] < details.lastConnect)
+				mapRsPgpId[details.gpg_id] = details.lastConnect;
+		}else{
```
Comment (G10h4ck) [inline libresapi/src/api/PeersHandler.cpp]: Please use Allman indent style http://retroshare.readthedocs.io/en/latest/developer/coding/#coding-style

**Review by G10h4ck — CHANGES_REQUESTED** [review]
Nice work, please do some changes as per comments so we can integrate it into the code easly

File: libresapi/src/api/PeersHandler.cpp (hunk 1 above)
Comment (myfingerhurt) [reply libresapi/src/api/PeersHandler.cpp:817]: I'm trying to replace the `/peers/${PGP_ID}/delete` eventually.
And both removing `pgp_id` and removing `peer_id` are so similar, I think of putting them together may be more simplified.

Well, I can split them into two in the next commit.

File: libresapi/src/api/PeersHandler.cpp (hunk 2 above)
Comment (myfingerhurt) [reply libresapi/src/api/PeersHandler.cpp:854]: Add Unix Time notes.

File: libresapi/src/api/PeersHandler.cpp (hunk 10 above)
Comment (myfingerhurt) [reply libresapi/src/api/PeersHandler.cpp]: Done.

File: libresapi/src/api/PeersHandler.cpp (hunk 9 above)
Comment (myfingerhurt) [reply libresapi/src/api/PeersHandler.cpp]: Done.

File: libresapi/src/api/PeersHandler.cpp (hunk 11 above)
Comment (myfingerhurt) [reply libresapi/src/api/PeersHandler.cpp]: Done.

**Comment by G10h4ck** [comment]
Plese just drop the unwanted commits, not revert them to avoid history bloating as this isn't in master branch yet history rewriting is not a problem

just do
```
git reset --hard 6cb4d36e00ff49e5a370dd3a60dfcf62e8fed015
git push -f
```

File: libresapi/src/api/PeersHandler.cpp (hunk 12 above)
Comment (myfingerhurt) [reply libresapi/src/api/PeersHandler.cpp]: Done.

File: libresapi/src/api/PeersHandler.cpp (hunk 8 above)
Comment (myfingerhurt) [reply libresapi/src/api/PeersHandler.cpp]: Done.

File: libresapi/src/api/PeersHandler.cpp (hunk 7 above)
Comment (myfingerhurt) [reply libresapi/src/api/PeersHandler.cpp]: Done.

File: libresapi/src/api/PeersHandler.cpp (hunk 6 above)
Comment (myfingerhurt) [reply libresapi/src/api/PeersHandler.cpp]: Done.

File: libresapi/src/api/PeersHandler.cpp (hunk 5 above)
Comment (myfingerhurt) [reply libresapi/src/api/PeersHandler.cpp]: Done.

File: libresapi/src/api/PeersHandler.cpp (hunk 4 above)
Comment (myfingerhurt) [reply libresapi/src/api/PeersHandler.cpp]: Done.

File: libresapi/src/api/PeersHandler.cpp (hunk 3 above)
Comment (myfingerhurt) [reply libresapi/src/api/PeersHandler.cpp]: Done.

**Comment by myfingerhurt** [comment]
I did what you had suggested.  Does it look better or worse?
The previous code seems not following the Allman indent style. I didn't touch that, but the copy and paste in notepad++ unified the indent style automatically.
```
git reset --hard xxxxxxxxxxxxxxxxx
git push -f
```

File: libresapi/src/api/PeersHandler.cpp (hunk 1 above)
Comment (G10h4ck) [reply libresapi/src/api/PeersHandler.cpp:817]: Please do that ;)

**Comment by G10h4ck** [comment]
Ok now please fix the Allman style thing ;)

**Comment by myfingerhurt** [comment]
Well, I believe I have fixed Allman style things with my last commit.
Is there anything I haven't done? I'm trying to make git to ignore leading spaces and tabs or generate a patch file without indentation, it seems none exists.

But with `?w=1`  it is better looking.

https://github.com/RetroShare/RetroShare/pull/1014/commits/f3485580773eaed3086d96219a11d21a3535a428?w=1

**Comment by G10h4ck** [comment]
The indentation fixes commit is not included in the PR probably it got caught in the revert/reset crossfire, you can cherry pick it probably and push again (no need for -f this time)

have you splitted PeersHandler::handleRemoveNode as requested?

**Comment by G10h4ck** [comment]
I have done the required modifications and merged it afterwards, thanks you for the contribution, please in the future use the qtcreator coding style file provided in the documentation so it is easier to merge your code ;)

**Pull request merged**